OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/backend/impl/InAppBackendService.kt

```diff
@@ -209,4 +201,88 @@ internal class InAppBackendService(
+            val values =
+                OptionalHeaders(
+                    offset = offset,
+                    secondsSinceAppOpen = sessionTime,
+                    retryCount = retryCount,
+                )
+            val response = _httpClient.get(baseUrl, values)
+
+            if (response.isSuccess) {
+                val jsonResponse = response.payload?.let { JSONObject(it) }
+                return jsonResponse?.let { hydrateInAppMessages(it) }
+            } else if (response.statusCode == 425) { // 425 Too Early
+                val retryLimit = response.retryLimit ?: DEFAULT_RETRY_LIMIT
+                val retryAfter = response.retryAfterSeconds ?: DEFAULT_RETRY_AFTER_SECONDS
+
+                if (attempts > retryLimit) {
+                    break
+                }
+
+                delay(retryAfter * 1_000L)
+            } else if (response.statusCode == 429) { // 429 Too Many Requests
+                val retryAfter = response.retryAfterSeconds ?: delayTime
+
+                if (attempts > (response.retryLimit ?: DEFAULT_RETRY_LIMIT)) {
+                    break
+                }
+
+                delay(retryAfter * 1_000L)
+                delayTime *= 2 // Exponential backoff
+            } else if (response.statusCode == 444 || response.statusCode >= 500) {
+                if (attempts > DEFAULT_RETRY_LIMIT) {
+                    break
+                }
+
+                delay(delayTime * 1_000L)
+                delayTime *= 2 // Exponential backoff
+            } else {
+                return null
+            }
+
+            attempts++
+        }
+
+        // If all retries fail, make a final attempt without the offset. This will tell the server,
+        // we give up, just give me the IAMs without first ensuring data consistency
+        return fetchInAppMessagesWithoutOffset(baseUrl, sessionTime)
+    }
+
+    private suspend fun fetchInAppMessagesWithoutOffset(
+        url: String,
+        sessionTime: Long,
+    ): List<InAppMessage>? {
+        val response = _httpClient.get(url, OptionalHeaders(secondsSinceAppOpen = sessionTime))
+
+        if (response.isSuccess) {
+            val jsonResponse = response.payload?.let { JSONObject(it) }
+            return jsonResponse?.let { hydrateInAppMessages(it) }
+        } else {
+            return null
+        }
+    }
+
+    private fun hydrateInAppMessages(jsonResponse: JSONObject): List<InAppMessage>? =
+        if (jsonResponse.has("in_app_messages")) {
+            val iamMessagesAsJSON = jsonResponse.getJSONArray("in_app_messages")
+            // TODO: Outstanding question on whether we still want to cache this.  Only used when
+            //       hard start of the app, but within 30 seconds of it being killed (i.e. same session startup).
+            // Cache copy for quick cold starts
+            // _prefs.savedIAMs = iamMessagesAsJSON.toString()
+            _hydrator.hydrateIAMMessages(iamMessagesAsJSON)
```

Comment:
I don't think we want to cache this any more, as it could lead to the wrong IAMs shown.